In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import pandas as pd

Folder Structure and File Format

In [3]:
import os
from collections import defaultdict

def count_file_types(folder):
    file_types = defaultdict(int)
    for file in os.listdir(folder):
        file_path = os.path.join(folder, file)
        if os.path.isfile(file_path):
            ext = os.path.splitext(file)[-1].lower().lstrip('.')
            file_types[ext] += 1
    return file_types

def list_folders(directory, indent=0):
    for item in os.listdir(directory):
        if item =='venv':
            continue
        path = os.path.join(directory, item)
        if os.path.isdir(path):
            print("  " * indent + f"📂 {item}")
            file_counts = count_file_types(path)
            if file_counts:
                file_counts_str = ", ".join(f"{count} {ext}" for ext, count in file_counts.items())
                print("  " * (indent + 1) + f"📄 {file_counts_str}")
            list_folders(path, indent + 1)

if __name__ == "__main__":
    current_directory = os.getcwd()
    print(f"Root: {current_directory}")
    list_folders(current_directory)

Root: d:\OneDrive - iitgn.ac.in\Desktop\Projects\Human-Activity-Recognition-using-Accelerometer-and-Gyroscope-Data
📂 .git
  📄 6 
  📂 hooks
    📄 14 sample
  📂 info
    📄 1 
  📂 logs
    📄 1 
    📂 refs
      📂 heads
        📄 1 
      📂 remotes
        📂 origin
          📄 2 
  📂 objects
    📂 03
      📄 1 
    📂 07
      📄 1 
    📂 09
      📄 1 
    📂 0d
      📄 1 
    📂 14
      📄 1 
    📂 19
      📄 1 
    📂 1e
      📄 1 
    📂 27
      📄 1 
    📂 29
      📄 2 
    📂 2b
      📄 1 
    📂 2c
      📄 1 
    📂 2e
      📄 1 
    📂 38
      📄 1 
    📂 43
      📄 1 
    📂 45
      📄 1 
    📂 50
      📄 1 
    📂 54
      📄 1 
    📂 63
      📄 1 
    📂 6f
      📄 1 
    📂 79
      📄 1 
    📂 7b
      📄 1 
    📂 81
      📄 1 
    📂 83
      📄 1 
    📂 8a
      📄 1 
    📂 8b
      📄 1 
    📂 8e
      📄 1 
    📂 91
      📄 1 
    📂 92
      📄 1 
    📂 93
      📄 2 
    📂 9e
      📄 1 
    📂 a3
      📄 2 
    📂 a5
      📄 1 
    📂 ae
      📄 2 
    📂 b2
      📄 1 
    📂 cb
      📄 1 
    📂 cd
  

## a.) Use UCI HAR dataset and build a model to predict various activity classes.

In [4]:
X_train = pd.read_csv("D:\\OneDrive - iitgn.ac.in\\Desktop\\Projects\\Human Activity Recognization AI_sensing for health_\\UCI HAR Dataset\\train\\X_train.txt", header=None, delim_whitespace=True)
y_train = pd.read_csv("D:\\OneDrive - iitgn.ac.in\Desktop\\Projects\\Human Activity Recognization AI_sensing for health_\\UCI HAR Dataset\\train\\y_train.txt", header = None)
x_test = pd.read_csv("D:\\OneDrive - iitgn.ac.in\Desktop\\Projects\\Human Activity Recognization AI_sensing for health_\\UCI HAR Dataset\\test\\X_test.txt", header= None, delim_whitespace = True)
y_test = pd.read_csv("D:\\OneDrive - iitgn.ac.in\Desktop\\Projects\\Human Activity Recognization AI_sensing for health_\\UCI HAR Dataset\\test\\y_test.txt", header = None)

C:\Users\Asus\AppData\Local\Temp\ipykernel_21960\4004002307.py:1: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  X_train = pd.read_csv("D:\\OneDrive - iitgn.ac.in\\Desktop\\Projects\\Human Activity Recognization AI_sensing for health_\\UCI HAR Dataset\\train\\X_train.txt", header=None, delim_whitespace=True)
C:\Users\Asus\AppData\Local\Temp\ipykernel_21960\4004002307.py:3: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  x_test = pd.read_csv("D:\\OneDrive - iitgn.ac.in\Desktop\\Projects\\Human Activity Recognization AI_sensing for health_\\UCI HAR Dataset\\test\\X_test.txt", header= None, delim_whitespace = True)


In [5]:
X_train.head()

0         1         2         3         4         5         6    \
0  0.288585 -0.020294 -0.132905 -0.995279 -0.983111 -0.913526 -0.995112   
1  0.278419 -0.016411 -0.123520 -0.998245 -0.975300 -0.960322 -0.998807   
2  0.279653 -0.019467 -0.113462 -0.995380 -0.967187 -0.978944 -0.996520   
3  0.279174 -0.026201 -0.123283 -0.996091 -0.983403 -0.990675 -0.997099   
4  0.276629 -0.016570 -0.115362 -0.998139 -0.980817 -0.990482 -0.998321   

        7         8         9    ...       551       552       553       554  \
0 -0.983185 -0.923527 -0.934724  ... -0.074323 -0.298676 -0.710304 -0.112754   
1 -0.974914 -0.957686 -0.943068  ...  0.158075 -0.595051 -0.861499  0.053477   
2 -0.963668 -0.977469 -0.938692  ...  0.414503 -0.390748 -0.760104 -0.118559   
3 -0.982750 -0.989302 -0.938692  ...  0.404573 -0.117290 -0.482845 -0.036788   
4 -0.979672 -0.990441 -0.942469  ...  0.087753 -0.351471 -0.699205  0.123320   

        555       556       557       558       559       560  
0  0.030400 -0.464761 -0.018446 -0.841247  0.179941 -0.058627  
1 -0.007435 -0.732626  0.703511 -0.844788  0.180289 -0.054317  
2  0.177899  0.100699  0.808529 -0.848933  0.180637 -0.049118  
3 -0.012892  0.640011 -0.485366 -0.848649  0.181935 -0.047663  
4  0.122542  0.693578 -0.615971 -0.847865  0.185151 -0.043892  

[5 rows x 561 columns]

In [6]:
y_train.head()

0
0  5
1  5
2  5
3  5
4  5

In [7]:
x_test.head()

0         1         2         3         4         5         6    \
0  0.257178 -0.023285 -0.014654 -0.938404 -0.920091 -0.667683 -0.952501   
1  0.286027 -0.013163 -0.119083 -0.975415 -0.967458 -0.944958 -0.986799   
2  0.275485 -0.026050 -0.118152 -0.993819 -0.969926 -0.962748 -0.994403   
3  0.270298 -0.032614 -0.117520 -0.994743 -0.973268 -0.967091 -0.995274   
4  0.274833 -0.027848 -0.129527 -0.993852 -0.967445 -0.978295 -0.994111   

        7         8         9    ...       551       552       553       554  \
0 -0.925249 -0.674302 -0.894088  ...  0.071645 -0.330370 -0.705974  0.006462   
1 -0.968401 -0.945823 -0.894088  ... -0.401189 -0.121845 -0.594944 -0.083495   
2 -0.970735 -0.963483 -0.939260  ...  0.062891 -0.190422 -0.640736 -0.034956   
3 -0.974471 -0.968897 -0.938610  ...  0.116695 -0.344418 -0.736124 -0.017067   
4 -0.965953 -0.977346 -0.938610  ... -0.121711 -0.534685 -0.846595 -0.002223   

        555       556       557       558       559       560  
0  0.162920 -0.825886  0.271151 -0.720009  0.276801 -0.057978  
1  0.017500 -0.434375  0.920593 -0.698091  0.281343 -0.083898  
2  0.202302  0.064103  0.145068 -0.702771  0.280083 -0.079346  
3  0.154438  0.340134  0.296407 -0.698954  0.284114 -0.077108  
4 -0.040046  0.736715 -0.118545 -0.692245  0.290722 -0.073857  

[5 rows x 561 columns]

In [8]:
y_test.head()

0
0  5
1  5
2  5
3  5
4  5

In [9]:
X_train.shape, y_train.shape, x_test.shape, y_test.shape

((7352, 561), (7352, 1), (2947, 561), (2947, 1))

In [14]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, f1_score, classification_report
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

### Scale the data for better fit

In [11]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

X_train=scaler.fit_transform(X_train)
X_test = scaler.transform(x_test)

### Model Training

In [12]:
# Function which task data and defined model as input then train the model and print accuracy and Classification Report

def Check_Model(model, X_train, y_train, X_test, y_test):
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)

    print(f"Train Accuracy Score is : {accuracy_score(model.predict(X_train), y_train)}")
    print(f"Test Accuracy Score is : {accuracy_score(y_pred, y_test)}")
    print(f"\n Classification Report\n {classification_report(y_pred, y_test)}")

#### Logistic Regression

In [87]:
Model1_LR = LogisticRegression()
Check_Model(Model1_LR, X_train, y_train, X_test, y_test)

Train Accuracy Score is : 0.9964635473340587
Test Accuracy Score is : 0.9545300305395318

 Classification Report
               precision    recall  f1-score   support

           1       0.99      0.94      0.97       522
           2       0.95      0.95      0.95       469
           3       0.93      0.99      0.96       396
           4       0.88      0.97      0.92       444
           5       0.97      0.89      0.93       582
           6       0.99      1.00      1.00       534

    accuracy                           0.95      2947
   macro avg       0.95      0.96      0.95      2947
weighted avg       0.96      0.95      0.95      2947



In [89]:
# Hyper Parameter Tunning
from  sklearn.model_selection import GridSearchCV

parameters_grid ={
    'C':[0.001,0.1, 10],
    'penalty':['l1','l2'],
    'solver' : ['liblinear', 'saga']
}

model = LogisticRegression()

# Perform GridSeaerchCV
grid_search = GridSearchCV(model, parameters_grid, cv = 5, scoring='accuracy', n_jobs=-1, verbose=1)
grid_search.fit(X_train, y_train)

# Best parameters and best score
print("Best Hyperparameters:", grid_search.best_params_)
print("Best Accuracy:", grid_search.best_score_)

best_model = grid_search.best_estimator_
y_pred = best_model.predict(X_test)

print("Test Accuracy Score:", accuracy_score(y_test, y_pred))
print("Classification Report:\n", classification_report(y_test, y_pred))

Fitting 5 folds for each of 12 candidates, totalling 60 fits
Best Hyperparameters: {'C': 10, 'penalty': 'l1', 'solver': 'liblinear'}
Best Accuracy: 0.9402965264963905
Test Accuracy Score: 0.9609772650152698
Classification Report:
               precision    recall  f1-score   support

           1       0.97      0.99      0.98       496
           2       0.96      0.97      0.97       471
           3       1.00      0.97      0.99       420
           4       0.97      0.85      0.91       491
           5       0.89      0.97      0.93       532
           6       0.99      1.00      1.00       537

    accuracy                           0.96      2947
   macro avg       0.96      0.96      0.96      2947
weighted avg       0.96      0.96      0.96      2947



#### Decision Tree

In [15]:
Model2 = DecisionTreeClassifier()
Check_Model(Model2, X_train, y_train, X_test, y_test)

Train Accuracy Score is : 1.0
Test Accuracy Score is : 0.8551068883610451

 Classification Report
               precision    recall  f1-score   support

           1       0.90      0.83      0.87       539
           2       0.76      0.81      0.79       440
           3       0.82      0.85      0.84       408
           4       0.76      0.83      0.80       451
           5       0.86      0.80      0.83       572
           6       1.00      1.00      1.00       537

    accuracy                           0.86      2947
   macro avg       0.85      0.85      0.85      2947
weighted avg       0.86      0.86      0.86      2947



### Random Forest

In [72]:
Model3 = RandomForestClassifier()
Check_Model(Model3, X_train, y_train, X_test, y_test)

Train Accuracy Score is : 1.0
Test Accuracy Score is : 0.9260264675941635

 Classification Report
               precision    recall  f1-score   support

           1       0.97      0.89      0.93       540
           2       0.92      0.90      0.91       478
           3       0.84      0.96      0.90       369
           4       0.88      0.91      0.90       476
           5       0.92      0.90      0.91       547
           6       1.00      1.00      1.00       537

    accuracy                           0.93      2947
   macro avg       0.92      0.93      0.92      2947
weighted avg       0.93      0.93      0.93      2947



#### Support Vetor Machines

In [73]:
from sklearn.svm import SVC

model4 = SVC()
Check_Model(model4, X_train, y_train, X_test, y_test)

Train Accuracy Score is : 0.9866702937976061
Test Accuracy Score is : 0.9521547336274178

 Classification Report
               precision    recall  f1-score   support

           1       0.97      0.96      0.97       502
           2       0.97      0.93      0.95       491
           3       0.92      0.98      0.95       395
           4       0.90      0.94      0.92       471
           5       0.95      0.92      0.93       549
           6       1.00      1.00      1.00       539

    accuracy                           0.95      2947
   macro avg       0.95      0.95      0.95      2947
weighted avg       0.95      0.95      0.95      2947



In [16]:
accuracy_author_data ={"SVC": 0.9521,
                       "Logistic Regression": 0.9609,
                       "RandomForestClassifier":0.9260}

The best model we got are Logistic Regression and Support Vector Classifier. We can improve accuracy by performing hyperparameter tunning of Decision tree classier, Random Forest and SVC but I coun't because of less computation

## b.) Use Deep Learning models like LSTM,1D Cnns for modeling using inertial sensors data (Raw accelerometer data). Do not train DL methods on already existing features provided by the authors

We will use tensorflow to train models

In [62]:
train_folder = "D:/OneDrive - iitgn.ac.in/Desktop/Projects/Human Activity Recognization AI_sensing for health_/UCI HAR Dataset/train/Inertial Signals"

for file in os.listdir(train_folder):
    print(file)


body_acc_x_train.txt
body_acc_y_train.txt
body_acc_z_train.txt
body_gyro_x_train.txt
body_gyro_y_train.txt
body_gyro_z_train.txt
total_acc_x_train.txt
total_acc_y_train.txt
total_acc_z_train.txt


#### Data loading and Preprocessing

In [63]:
base_dir = "D:/OneDrive - iitgn.ac.in/Desktop/Projects/Human Activity Recognization AI_sensing for health_/UCI HAR Dataset"

# defining function to load raw sensor data
def load_raw_sensor_data(folder, signals):
    data = []
    for signal in signals:
        file_path = os.path.join(folder, f"{signal}.txt")
        sensor_data = pd.read_csv(file_path, delim_whitespace=True, header=None).values
        data.append(sensor_data) 
    return np.stack(data, axis=-1)  

# List of inertial sensor signals (accererometer + gyroscope)
sensor_signals_train = ["body_acc_x_train", "body_acc_y_train", "body_acc_z_train",
                  "body_gyro_x_train", "body_gyro_y_train", "body_gyro_z_train"]
sensor_signals_test = ["body_acc_x_test", "body_acc_y_test", "body_acc_z_test",
                  "body_gyro_x_test", "body_gyro_y_test", "body_gyro_z_test"]

# Load train and test data
train_folder = os.path.join(base_dir, "train", "Inertial Signals")
test_folder = os.path.join(base_dir, "test", "Inertial Signals")

In [64]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv1D, LSTM, Dense, Dropout, Flatten, BatchNormalization, MaxPooling1D
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, StandardScaler

In [65]:
# Load train and test data
X_train = load_raw_sensor_data(train_folder, sensor_signals_train)
X_test = load_raw_sensor_data(test_folder, sensor_signals_test)

# Load activity labels
y_train = np.loadtxt(os.path.join(base_dir, "train", "y_train.txt")).astype(int)
y_test = np.loadtxt(os.path.join(base_dir, "test", "y_test.txt")).astype(int)


C:\Users\Asus\AppData\Local\Temp\ipykernel_12664\2803395076.py:8: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  sensor_data = pd.read_csv(file_path, delim_whitespace=True, header=None).values
C:\Users\Asus\AppData\Local\Temp\ipykernel_12664\2803395076.py:8: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  sensor_data = pd.read_csv(file_path, delim_whitespace=True, header=None).values
C:\Users\Asus\AppData\Local\Temp\ipykernel_12664\2803395076.py:8: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  sensor_data = pd.read_csv(file_path, delim_whitespace=True, header=None).values
C:\Users\Asus\AppData\Local\Temp\ipykernel_12664\2803395076.py:8: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is 

In [66]:
# One-hot encode labels
num_classes = len(np.unique(y_train))
y_train = keras.utils.to_categorical(y_train - 1, num_classes)
y_test = keras.utils.to_categorical(y_test - 1, num_classes)

In [67]:
# Normalization
scaler = StandardScaler()
X_train = np.reshape(X_train, (-1, X_train.shape[-1]))  # Flatten before scaling the data
X_test = np.reshape(X_test, (-1, X_test.shape[-1])) 

In [68]:
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [69]:
print(f"X_train shape: {X_train.shape}")  # Should be (samples, 128, 9)
print(f"y_train shape: {y_train.shape}")  # Should be (samples, num_classes)


X_train shape: (941056, 6)
y_train shape: (7352, 6)


In [70]:
import numpy as np

num_samples = (X_train.shape[0] // 128) * 128  # Make it a multiple of 128
X_train = X_train[:num_samples]
X_train = np.reshape(X_train, (-1, 128, 6))  # Use 6 instead of 9

num_samples_test = (X_test.shape[0] // 128) * 128
X_test = X_test[:num_samples_test]
X_test = np.reshape(X_test, (-1, 128, 6))  # Use 6 instead of 9

print(X_train.shape)  # Expected: (some_batch_size, 128, 6)
print(X_test.shape)   # Expected: (some_batch_size, 128, 6)


(7352, 128, 6)
(2947, 128, 6)


#### Model Building and Training

In [75]:
# 1D CNN Model
def create_cnn_model():
    model = Sequential([
        Conv1D(64, kernel_size=3, activation='relu', input_shape=(128, 6)),
        BatchNormalization(),
        MaxPooling1D(pool_size=2),
        Conv1D(128, kernel_size=3, activation='relu'),
        BatchNormalization(),
        MaxPooling1D(pool_size=2),
        Flatten(),
        Dense(128, activation='relu'),
        Dropout(0.5),
        Dense(num_classes, activation='softmax')
    ])
    model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
    return model

In [76]:
# LSTM Model
def create_lstm_model():
    model = Sequential([
        LSTM(64, return_sequences=True, input_shape=(128, 6)),
        LSTM(128),
        Dense(128, activation='relu'),
        Dropout(0.5),
        Dense(num_classes, activation='softmax')
    ])
    model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
    return model

In [78]:
# Define Models
cnn_model = create_cnn_model()
lstm_model = create_lstm_model()

In [79]:
# train CNN model
cnn_model.fit(X_train, y_train, epochs=20, batch_size=32, validation_data=(X_test, y_test))

Epoch 1/20
230/230 ━━━━━━━━━━━━━━━━━━━━ 5s 12ms/step - accuracy: 0.5076 - loss: 1.5983 - val_accuracy: 0.6549 - val_loss: 0.7100
Epoch 2/20
230/230 ━━━━━━━━━━━━━━━━━━━━ 3s 14ms/step - accuracy: 0.6829 - loss: 0.6287 - val_accuracy: 0.6753 - val_loss: 0.8821
Epoch 3/20
230/230 ━━━━━━━━━━━━━━━━━━━━ 3s 13ms/step - accuracy: 0.7044 - loss: 0.5794 - val_accuracy: 0.7584 - val_loss: 0.6204
Epoch 4/20
230/230 ━━━━━━━━━━━━━━━━━━━━ 4s 15ms/step - accuracy: 0.7421 - loss: 0.5267 - val_accuracy: 0.7879 - val_loss: 0.5643
Epoch 5/20
230/230 ━━━━━━━━━━━━━━━━━━━━ 4s 18ms/step - accuracy: 0.7648 - loss: 0.4814 - val_accuracy: 0.7655 - val_loss: 0.5491
Epoch 6/20
230/230 ━━━━━━━━━━━━━━━━━━━━ 3s 14ms/step - accuracy: 0.7550 - loss: 0.4915 - val_accuracy: 0.7937 - val_loss: 0.4703
Epoch 7/20
230/230 ━━━━━━━━━━━━━━━━━━━━ 3s 13ms/step - accuracy: 0.7868 - loss: 0.4282 - val_accuracy: 0.7954 - val_loss: 0.5044
Epoch 8/20
230/230 ━━━━━━━━━━━━━━━━━━━━ 3s 13ms/step - accuracy: 0.7957 - loss: 0.4275 - val_accu

In [80]:
cnn_acc = cnn_model.evaluate(X_test, y_test)[1]

93/93 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8767 - loss: 0.4615


In [81]:
# Train LSTM model
lstm_model.fit(X_train, y_train, epochs=20, batch_size=32, validation_data=(X_test, y_test))

Epoch 1/20
230/230 ━━━━━━━━━━━━━━━━━━━━ 22s 86ms/step - accuracy: 0.3935 - loss: 1.2796 - val_accuracy: 0.5914 - val_loss: 0.7914
Epoch 2/20
230/230 ━━━━━━━━━━━━━━━━━━━━ 19s 82ms/step - accuracy: 0.6146 - loss: 0.6874 - val_accuracy: 0.5589 - val_loss: 0.9968
Epoch 3/20
230/230 ━━━━━━━━━━━━━━━━━━━━ 19s 82ms/step - accuracy: 0.6165 - loss: 0.6798 - val_accuracy: 0.6128 - val_loss: 0.7439
Epoch 4/20
230/230 ━━━━━━━━━━━━━━━━━━━━ 22s 97ms/step - accuracy: 0.6254 - loss: 0.6573 - val_accuracy: 0.6247 - val_loss: 0.7162
Epoch 5/20
230/230 ━━━━━━━━━━━━━━━━━━━━ 23s 99ms/step - accuracy: 0.6284 - loss: 0.6467 - val_accuracy: 0.6308 - val_loss: 0.6939
Epoch 6/20
230/230 ━━━━━━━━━━━━━━━━━━━━ 16s 69ms/step - accuracy: 0.6397 - loss: 0.6339 - val_accuracy: 0.6132 - val_loss: 0.7025
Epoch 7/20
230/230 ━━━━━━━━━━━━━━━━━━━━ 18s 79ms/step - accuracy: 0.6368 - loss: 0.6141 - val_accuracy: 0.6301 - val_loss: 0.7218
Epoch 8/20
230/230 ━━━━━━━━━━━━━━━━━━━━ 18s 77ms/step - accuracy: 0.6397 - loss: 0.6142 - 

In [83]:
lstm_acc = lstm_model.evaluate(X_test, y_test)[1]

93/93 ━━━━━━━━━━━━━━━━━━━━ 2s 23ms/step - accuracy: 0.6070 - loss: 0.7324


In [84]:
# Accuracies
print(f"1D CNN Accuracy: {cnn_acc * 100:.2f}%")
print(f"LSTM Accuracy: {lstm_acc * 100:.2f}%")

1D CNN Accuracy: 88.26%
LSTM Accuracy: 62.95%


CNN is giving more accuracy

## c.) How can the same be done using Machine Learning models like Random forest, SVM, Logistic regression. Use TSFEL library to generate features from Inertial data. Compare the performance of models trained on your generated features versus features provided by the authors.

In [17]:
import tsfel
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [39]:
# Extracting Features using TSFEL
cfg = tsfel.get_features_by_domain("all")

In [38]:
X_train.shape[0],X_test.shape[0]

(7352, 2947)

In [ ]:
X_train_features = [tsfel.time_series_features_extractor(cfg, X_train[i], verbose=0) for i in range(X_train.shape[0])]

In [ ]:
X_test_features = [tsfel.time_series_features_extractor(cfg, X_test[i], verbose=0) for i in range(X_test.shape[0])]

Note: I cleared step shown as output of TSFEL feature extracted to enhance readability

In [52]:
np.array(X_train_features).shape, np.array(X_test_features).shape

((7352, 1, 162), (2947, 1, 162))

here X_train_features and X_test_features has an extra dimention. I need to remove that in order to convert into dataframe and further tranning

In [53]:
X_train_features = np.array(X_train_features).squeeze()  # Remove the extra dimension
X_test_features = np.array(X_test_features).squeeze()  # Remove the extra dimension

# Convert to DataFrame
X_train_df = pd.DataFrame(X_train_features)
X_test_df = pd.DataFrame(X_test_features)

# Check, is there any missing values
print(X_train_df.isnull().sum().sum())  
print(X_test_df.isnull().sum().sum())


0
0


No missing values

In [54]:
X_train_df.shape, X_test_df.shape

((7352, 162), (2947, 162))

In [55]:
accuracy_tsfel = dict()

#### Train Random Forest Model

In [56]:
type(y_train)

pandas.core.frame.DataFrame

In [59]:
model = RandomForestClassifier(n_estimators=25, random_state=42)
model.fit(X_train_features, y_train)

d:\OneDrive - iitgn.ac.in\Desktop\Projects\Human Activity Recognization AI_sensing for health_\ve\lib\site-packages\sklearn\base.py:1389: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


RandomForestClassifier(n_estimators=25, random_state=42)

In [60]:
accuracy = accuracy_score(y_test, model.predict(X_test_features))
print(f"Random Forest Accuracy: {accuracy * 100:.2f}%")
accuracy_tsfel["RandomForestClassifier"] = accuracy

Random Forest Accuracy: 77.13%


#### Train SVM Model

In [61]:
model = SVC(kernel='rbf', probability=True)
model.fit(X_train_features, y_train)

d:\OneDrive - iitgn.ac.in\Desktop\Projects\Human Activity Recognization AI_sensing for health_\ve\lib\site-packages\sklearn\utils\validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVC(probability=True)

In [62]:
accuracy = accuracy_score(y_test, model.predict(X_test_features))
print(f"SVC Accuracy: {accuracy * 100:.2f}%")
accuracy_tsfel["SVC"] = accuracy

SVC Accuracy: 29.11%


#### Train Logistic Regression Model

In [63]:
model = LogisticRegression(max_iter=1000)
model.fit(X_train_features, y_train)

d:\OneDrive - iitgn.ac.in\Desktop\Projects\Human Activity Recognization AI_sensing for health_\ve\lib\site-packages\sklearn\utils\validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
d:\OneDrive - iitgn.ac.in\Desktop\Projects\Human Activity Recognization AI_sensing for health_\ve\lib\site-packages\sklearn\linear_model\_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(max_iter=1000)

In [64]:
accuracy = accuracy_score(y_test, model.predict(X_test_features))
print(f"Logistic Regression Accuracy: {accuracy * 100:.2f}%")
accuracy_tsfel["Logistic Regression"] = accuracy

Logistic Regression Accuracy: 61.86%


#### Compare the performance of models trained on your generated features versus features provided by the authors

In [65]:
print("Accuracy of models on author's data:", accuracy_author_data)
print("\n\nAccuracy of models on TSFEL extracted features:", accuracy_tsfel)

Accuracy of models on author's data: {'SVC': 0.9521, 'Logistic Regression': 0.9609, 'RandomForestClassifier': 0.926}


Accuracy of models on TSFEL extracted features: {'RandomForestClassifier': 0.7712928401764506, 'SVC': 0.2911435357991177, 'Logistic Regression': 0.6185951815405497}


## Create a short summary explaining the approach and observations. 

Approach and Observation summary 
- This project aimed to classify human activities using the UCI HAR dataset, comparing traditional machine learning and deep learning approaches. At first I preprocessed the data given by author by Standardizing. Then I trained Logistic Regression (96.09% accuracy), Decision Tree (85.51% accuracy), random forest classifier (92.60) and Support vector classifier (95.21). 
- In second part I applied Deep learning models, such as 1D CNN and LSTM, where CNN outperformed LSTM (88.26% vs. 62.95%).
 - In Third part Feature extraction using TSFEL was explored, but results showed a performance drop, which shows that the features of the original dataset are more effective. Machine Learning Models such as SVC, Logistic Regression and RandomforestClassifier, all performed better on feature given by author than features extracted by TSFEL.
 - For performance drop, There are few reasons might responsible such as : I used only 6 row features where 3 axis of accelerometer and 3 axis of gyroscope, I avoided total acceleration features due to limited computation, this might cause loss of information and second reason could be author might used more complex feature extraction techniques.
